### PR TITLE
add documentation

### PR DESCRIPTION
**Pull request by jialuw96**

### CODE REVIEW

**Comment by jialuw96** [comment]
@adowling2 Hi Alex, please see index.rst for the documentation. I also sent you a pdf version through email. Thank you!

**Comment by jialuw96** [comment]
@adowling2 This is the documentation pr

File: pyomo/contrib/doe/documentation/docs/source/usage.rst [hunk 1]
```diff
@@ -0,0 +1,34 @@
+Usage
+=====
+
+.. _installation:
+
+Installation
+------------
+
+To use Lumache, first install it using pip:
```
Comment (adowling2) [inline pyomo/contrib/doe/documentation/docs/source/usage.rst:9]: Can this file be removed? Is it needed?

File: pyomo/contrib/doe/documentation/docs/source/index.rst [hunk 2]
```diff
@@ -0,0 +1,453 @@
+############
+Pyomo.DOE
+############
+
+Pyomo.DOE
+===================================
+
+**Pyomo.DOE** (Pyomo Design of Experiments) is a Python library for model-based design of experiments
+using science-based models.
```
Comment (adowling2) [inline pyomo/contrib/doe/documentation/docs/source/index.rst:9]: Remove extra line break here.

File: pyomo/contrib/doe/documentation/docs/source/index.rst [hunk 3]
```diff
@@ -0,0 +1,453 @@
...
+
+It was developed by **Jialu Wang** and **Alexander W. Dowling** of University of Notre Dame as part of the Carbon Capture Simulation for Industry Impact (CCSI2)
```
Comment (adowling2) [inline pyomo/contrib/doe/documentation/docs/source/index.rst:11]: Replace "It" with "Pyomo.DOE"

File: pyomo/contrib/doe/documentation/docs/source/index.rst [hunk 4]
```diff
@@ -0,0 +1,453 @@
...
+(https://www.acceleratecarboncapture.org/ ) project, 
+funded through the U.S. DOE Office of Fossil Energy by the Lawrence Berkeley National Laboratory through contract
```
Comment (adowling2) [inline pyomo/contrib/doe/documentation/docs/source/index.rst:13]: You can remove the "by the ..." and end the sentence after Energy.

File: pyomo/contrib/doe/documentation/docs/source/index.rst [hunk 5]
```diff
@@ -0,0 +1,453 @@
...
+
+If you use Pyomo.DOE, please cite:
+
+[Wang and Dowling, 2022] Jialu Wang, Alexander Dowling (2022), Pyomo.DOE: An Open-Source Package for Model-Based Design of Experiments in Python. (under review)
+
+Methodology Overview
+---------------------
+
+Model-based Design of Experiments (MBDoE) is a technique to maximize the information gain of experiments by directly using
+science-based models with physically meaningful parameters. It is one key component in the model calibration and uncertainty quantification workflow
```
Comment (adowling2) [inline pyomo/contrib/doe/documentation/docs/source/index.rst:24]: Remove extra line break.

File: pyomo/contrib/doe/documentation/docs/source/index.rst [hunk 6]
```diff
@@ -0,0 +1,453 @@
...
+shown below:
+
+.. figure:: flowchart.png
+   :scale: 25 %
+   :alt: map to buried treasure
+
+   The exploratory analysis, parameter estimation, uncertainty analysis, and MBDoE are combined into an iterative framework to select, refine, and calibrate science-based mathematical models with quantified uncertainty. The goal of this workflow is to achieve a confident parameter estimation.
```
Comment (adowling2) [inline pyomo/contrib/doe/documentation/docs/source/index.rst:31]: Replace "The goal... parameter estimation" with Currently, Pyomo.DOE focused on increasing parameter precision.

File: pyomo/contrib/doe/documentation/docs/source/index.rst [hunk 7]
```diff
@@ -0,0 +1,453 @@
...
+
+Pyomo.DOE provides the exploratory anlaysis and MBDoE capabilities to the Pyomo ecosystem. The user provides one Pyomo model, a set of parameter nominal values,
+the allowable design spaces for design variables, and the assumed observation error model.
+During exploratory analysis, Pyomo.DOE checks if this model parameters can be inferred from the postulated measurements or preliminary data.
+MBDoE then recommends optimized experimental conditions for collecting more data.
+Parameter estimation packages such as **Parmest** can perform parameter estimation using the available data to infer values for parameters,
+and facilitate an uncertainty analysis to approximate the parameter covariance matrix.
+If the parameter uncertainties are sufficiently small, the workflow terminates and returns the final model with quantified parametric uncertainty.
+If not, MBDoE is performed to recommend new experiments to generate new data.
```
Comment (adowling2) [inline pyomo/contrib/doe/documentation/docs/source/index.rst:40]: Replace last sentence: If not, MBDOE recommends optimized experimental conditions to generate new data.

File: pyomo/contrib/doe/documentation/docs/source/index.rst [hunk 8]
```diff
@@ -0,0 +1,453 @@
...
+
+Below is an overview of the type of optimization models Pyomo.DOE can accomodate:
+
+* Pyomo.DOE is suitable for optimization models of **continuous** variables
+* Pyomo.DOE can handle **equality constraints** defining state variables
+* Pyomo.DOE allows for (Partial) Differential-Algebraic Equations (PDAE) models which can be discretized and solved by Pyomo.DAE
```
Comment (adowling2) [inline pyomo/contrib/doe/documentation/docs/source/index.rst:46]: Pyomo.DOE supports (Partial) Differential Algebraic Equation (PDAE) models via Pyomo.DAE

File: pyomo/contrib/doe/documentation/docs/source/index.rst [hunk 9]
```diff
@@ -0,0 +1,453 @@
...
+* Pyomo.DOE also supports only algebraic models
```
Comment (adowling2) [inline pyomo/contrib/doe/documentation/docs/source/index.rst:47]: Pyomo.DOE also supports models with only algebraic constraints

File: pyomo/contrib/doe/documentation/docs/source/index.rst [hunk 10]
```diff
@@ -0,0 +1,453 @@
...
+*  :math:`\overline{\mathbf{w}} \in \mathbb{R}^{N_w}` are time-invariant decision variables.
+*  :math:`\mathbf{y} \subseteq \mathcal{Y}` are measurement response variables,   :math:`\mathcal{Y} \in \mathbb{R}^{N_r \times N_t}`.
+*  :math:`\mathbf{f}(\cdot)` are differential equations.
+*  :math:`\mathbf{g}(\cdot)` are algebraic equations.
+*  :math:`\mathbf{h}(\cdot)` are measurement functions.
+*  :math:`\mathbf{t} \in \mathbb{R}^{N_t \times 1}` is a union of all time sets.
+
+.. note::
+    * Process models provided to Pyomo.DOE should define an extra scenario index for all state variables and all parameters, as the first index before any other index.
+    * Process models define time as t. For steady-state models, t should be [0].
```
Comment (adowling2) [inline pyomo/contrib/doe/documentation/docs/source/index.rst:76]: Format t to appear as code. I believe this is `t` in markdown.

Comment (adowling2) [inline pyomo/contrib/doe/documentation/docs/source/index.rst:76]: Process models must include an index for time, named `t`.

File: pyomo/contrib/doe/documentation/docs/source/index.rst [hunk 11]
```diff
@@ -0,0 +1,453 @@
...
+
+Pyomo.DOE Required Inputs
+--------------------------------
+The required inputs to the Pyomo.DOE solver are the following:
+
+* A function that creates the process model
+* Dictionary of parameters and their nominal value
+* Dictionary of measurements and their measurement time points
+* Dictionary of design variables and their control time points
+* Prior FIM
```
Comment (adowling2) [inline pyomo/contrib/doe/documentation/docs/source/index.rst:137]: What data type (dictionary, numpy array) should the Prior FIM be defined as? Did we define FIM earlier in the documentation?

File: pyomo/contrib/doe/documentation/docs/source/index.rst [hunk 12]
```diff
@@ -0,0 +1,453 @@
...
+* Subordinate local and global NLP optimization solver
```
Comment (adowling2) [inline pyomo/contrib/doe/documentation/docs/source/index.rst:138]: Remove "subordinate"

Comment (adowling2) [reply pyomo/contrib/doe/documentation/docs/source/index.rst:138]: Did we define NLP earlier in the document?

File: pyomo/contrib/doe/documentation/docs/source/index.rst [hunk 13]
```diff
@@ -0,0 +1,453 @@
...
+
+measurement_object : ``object``
+    A ``object`` of the measurements, provided by the measurement class.
+
+
+create_model : ``function``
+    A ``function`` representing a deterministic process model.
+
+
+prior_FIM : ``list``
```
Comment (adowling2) [inline pyomo/contrib/doe/documentation/docs/source/index.rst:156]: How is the matrix represented as a list?

File: pyomo/contrib/doe/documentation/docs/source/index.rst [hunk 14]
```diff
@@ -0,0 +1,453 @@
...
+   :scale: 25 %
+   :alt: map to buried treasure
+
+
+.. autoclass:: fim_doe.DesignOfExperiments
+    :members:  __init__, optimize_doe, compute_FIM, run_grid_search
+
+
+.. Note::
+    Optimize_doe() includes the following steps:
```
Comment (adowling2) [inline pyomo/contrib/doe/documentation/docs/source/index.rst:172]: Check capitalization. Should it be `optimize_doe()`? Also, format this as code.

File: pyomo/contrib/doe/documentation/docs/source/index.rst [hunk 15]
```diff
@@ -0,0 +1,453 @@
...
+
+.. autoclass:: fim_doe.Grid_Search_Result
+    :special-members: __init__
+
+
+
+Pyomo.DOE Usage Example
+------------------------
+
+We will use an example to illustrate the usage of Pyomo.DOE to a reaction kinetics example.
```
Comment (adowling2) [inline pyomo/contrib/doe/documentation/docs/source/index.rst:194]: We illustrate the usage of Pyomo.DOE using a reaction kinetics example (Wang and Dowling, 2022).

File: pyomo/contrib/doe/documentation/docs/source/index.rst [hunk 16]
```diff
@@ -0,0 +1,453 @@
...
+:math:`C_A(t), C_B(t), C_C(t)` are the time-varying concentrations of the species A, B, C, respectively.
+:math:`k_1, k_2` are the rates for the two chemical reactions using an Arrhenius equation with activation energies :math:`E_1, E_2` and pre-exponential factors :math:`A_1, A_2`.
+The goal of MBDoE is to optimize the experiment design variables :math:`\boldsymbol{\varphi} = (C_{A0}, T(t))`, where :math:`C_{A0},T(t)` are the initial concentration of species A and the time-varying reactor temperature, to maximize the precision of unknown model parameters :math:`\boldsymbol{\theta} = (A_1, E_1, A_2, E_2)` by measuring :math:`\mathbf{y}(t)=(C_A(t), C_B(t), C_C(t))`.
+The observation errors are assumed to be independent both in time and across measurements with a constant standard deviation of 1 M for each species.
+
+
+Step 0: Import Pyomo and the Pyomo.DOE module
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In anticipation of using the Pyomo.DOE and building the Pyomo process model:
```
Comment (adowling2) [inline pyomo/contrib/doe/documentation/docs/source/index.rst:222]: You can remove this line.

File: pyomo/contrib/doe/documentation/docs/source/index.rst [hunk 17]
```diff
@@ -0,0 +1,453 @@
...
+
+Step 1: Define the Pyomo process model
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The process model for the reaction kinetics problem is shown below.
+
+.. doctest::
+
+    >>> # === Create model ==
+    >>> m = pyo.ConcreteModel()
```
Comment (adowling2) [inline pyomo/contrib/doe/documentation/docs/source/index.rst:239]: I am confused? Shouldn't this define `create_model`?

File: pyomo/contrib/doe/documentation/docs/source/index.rst [hunk 18]
```diff
@@ -0,0 +1,453 @@
...
+    >>> parameter_dict = {'A1': 84.79, 'A2': 371.72, 'E1': 7.78, 'E2': 15.05}
+
+    >>> # === Define prior information ==
+    >>> prior_none = np.zeros((4,4))
+
+
+Step 3: Compute the FIM of a square MBDoE problem
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+This method computes an MBDoE optimization problem with no Degree of Freedom.
```
Comment (adowling2) [inline pyomo/contrib/doe/documentation/docs/source/index.rst:343]: degrees of freedom can be lower case

File: pyomo/contrib/doe/documentation/docs/source/index.rst [hunk 19]
```diff
@@ -0,0 +1,453 @@
...
+
+Step 4: Exploratory analysis (Enumeration)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Exploratory analysis is suggested to enumerate the design space to check if the problem is identifiable, i.e., ensure that D-, E-optimality metrics is not an infinite small number near zero, and Modified E-optimality is not a big number.
+
+Pyomo.DOE accomplishes the exploratory analysis with the ``run_grid_search`` function.
+It allows to define any number of design decisions. Heatmaps can be drawn by two design variables, fixing other design variables.
+1D curve can be drawn by one design variable, fixing all other variables.
+The function ``run_grid_search`` enumerates over the design space, each MBDoE problem accomplished by ``compute_FIM`` method. Therefore, ``run_grid_search`` can also choose from the two modes, ``sequential_finite`` and ``direct_kaug``.
```
Comment (adowling2) [inline pyomo/contrib/doe/documentation/docs/source/index.rst:387]: Therefore, ``run_grid_search`` supports only two modes: ``sequential_finite`` and ``direct_kaug``

File: pyomo/contrib/doe/documentation/docs/source/index.rst [hunk 20]
```diff
@@ -0,0 +1,453 @@
...
+.. figure:: grid-1.png
+   :scale: 35 %
+   :alt: map to buried treasure
+
+
+Step 5: Gradient-based optimization
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Pyomo.DOE accomplishes gradient-based optimization with ``optimize_doe`` function for A- and D-optimality design. It is
+strongly suggested that when using D-optimality the identifiability of the problems is firstly analyzed by ``run_grid_search``.
```
Comment (adowling2) [inline pyomo/contrib/doe/documentation/docs/source/index.rst:425]: This sentence does not make sense.

**Review by adowling2 — COMMENTED** [review]
@jialuw96 Here are some minor comments. You should merge this soon and then get help from the Pyomo or CCSI teams on running the doctest.

File: pyomo/contrib/doe/documentation/docs/source/index.rst (hunk 13 above)
Comment (jialuw96) [reply pyomo/contrib/doe/documentation/docs/source/index.rst:156]: checked. it is numpy array.

**Comment by jialuw96** [comment]
@adowling2 index.rst is ready for review. Thank you!

**Pull request merged**